[feat] #2 Oauth2 로그인 authentication을 구현하였습니다.

suhyeon7497 opened the pull request:
### ✨ 작업 내용
- authentication 부분을 구현하였습니다.
---

### ✨ 참고 사항
- 현재  /hello에 관한 authentication을 진행하고 있어 다른 부분은 잘 작동할 것입니다.
- CustomOAuth2UserService에 provider를 kakao로 입력시켜줘야 하는 부분이 많은 데 그 부분을 어떻게 refactoring 하면 좋을지 첨언 부탁드립니다.
---

### ⏰ 현재 버그
- yml 파일이 없어 /hello로 접속하면 오류가 발생합니다.
- (있을 땐 잘 됐습니다.)
- 차후 환경변수를 추가할 예정입니다.
---

### ✏ Git Close

suhyeon7497 committed:
chore: security 관련 dependency 추가

oauth, security와 관련된 dependency를 추가하였습니다.

관련 이슈: #2

suhyeon7497 committed:
feat: user 정보 저장을 위한 요소 추가

UserEntity와 Repository를 추가 하였습니다.

관련 이슈: #2

suhyeon7497 committed:
feat: oauth 로그인을 위한 security config 설정

oauth 로그인을 위해
csrf를 차단 (rest api이기 때문에 jwt로 검증을 함),
formlogin을 차단(사용하지 않음),
httpBasic을 차단(사용하지 않음) 하였습니다.
session에 stateless 속성을 부여하였습니다.

관련 이슈: #2

suhyeon7497 committed:
chore: 민감한 환경변수를 위한 서브 모듈 추가

서브 모듈을 추가하여 민감한 정보를 관리합니다.

관련 이슈:#2

suhyeon7497 committed:
chore: 서브모듈 최신화

서브 모듈을 최신화 하여
api 키를 업데이트하였습니다

관련 이슈: #2

suhyeon7497 committed:
chore: application.properties 인식 문제 해결

.gitignore 파일에 application.properties를 넣지안았는데도
추적되지 않음. 그 문제를 해결

관련 이슈: #2

suhyeon7497 committed:
feat: authentication 구현

authentication을 구현하였습니다.
kakao만 한다는 가정하에 kakao를 직접 넣었습니다.

관련 이슈: #2

suhyeon7497 committed:
fix: authentication 하면 무한 리다이렉트 해결

/login 에 authenicated가 적용되어 있어 무한 리다이렉트 되던 문제를
해결하였습니다.

관련 이슈:#2

suhyeon7497 committed:
feat: oauth 로그인 하면 유저 정보 저장

oauth 로그인 시 유저 정보를 db에 저장

관련 이슈: #2

suhyeon7497 committed:
[chore]submodule 제거

submodule이 내 private repository라서 사용하지 않기로 정함
고로 삭제

관련 이슈: #2

suhyeon7497 committed:
Merge branch 'weekly/1' of https://github.com/kakao-tech-campus-2nd-step3/Team7_BE into feat/#2-authentication

suhyeon7497 committed:
[chore] 패키지 구조 변경

9/23일자로 변경된 패키지 컨벤션으로 패키지 구조를 변경하였습니다.

관련 이슈: #2

suhyeon7497 committed:
[fix] security 설정 없으면 오류 발생 해결

security에 대한 설정이 없으면 애플리케이션 실행 안돼서
application.yml 파일 추가

관련 이슈: #2

sanghee0820 commented on src/main/java/team7/inplace/security/application/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,47 @@
+
+@Service
+@RequiredArgsConstructor
+public class CustomOAuth2UserService extends DefaultOAuth2UserService {
```
DefaultOAuth2UserService를 Extends 하면 외부 API에 대한 의존성이 생겨 로직을 테스트 하기 쉽지 않습니다!
DefaultOAuth2UserService를 Bean으로 등록 한 후 주입받는 방법으로 수정해보시면 어떨까요?

추가로 Security에 관련된 설정의 경우에는 Lombok보다는, @Bean으로 직접 등록해주시는게 조금 더 좋아보입니다.
예를들어 DefaultOAuth2User은 Security Class라 @Service 등록이 불가능하니 직접 @Bean으로 등록해주어야합니다.
그런데 CustomOAuth2UserService는 Lombok으로 등록하고 DefaultOAuth2User는 @Bean으로 등록하면 일관성이 없어질 것 같네요!

sanghee0820 commented on src/main/java/team7/inplace/security/application/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,47 @@
+    public OAuth2User loadUser(OAuth2UserRequest oAuth2UserRequest)
+        throws OAuth2AuthenticationException {
+        String provider = oAuth2UserRequest.getClientRegistration().getRegistrationId();
+        if (!provider.equals(UserType.KAKAO.getProvider())) {
+            return null;
+        }
```
이 유효성 검사가 꼭 필요한 검사인가요?
이유를 듣고싶습니다!

suhyeon7497 replied:
원래는 다른 oauth로그인 까지 고려하였을때,
if(provider.equals(UserType.KAKAO.getProvider())) {
}
if(provider.equals(UserType.GOOGLE.getProvider())) {
}
를 적용하려 하였는데, google 로그인은 아직 사용하지 않아서 바꾸려다가 early return을 사용하려다 발생한 문제인 것 같습니다. 수정하도록 하겠습니다.

sanghee0820 commented on src/main/java/team7/inplace/security/application/dto/CustomOAuth2User.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import team7.inplace.security.domain.UserType;
+
+public record CustomOAuth2User(String username,
+                               String nickname,
+                               UserType userType) implements OAuth2User {
```
제가 보통 사용하는 선언 방식이랑 살짝 다른데 이렇게 선언하는건 어떠신지 의견을 듣고싶습니다.
```
public record CustomOAuth2User(
          String username,
          String nickname,
          UserType userType
) implements OAuth2User { }
```

suhyeon7497 replied:
넵 코드 컨벤션 지키도록 하겠습니다!

sanghee0820 commented on src/main/java/team7/inplace/security/application/dto/KakaoOAuthResponse.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.Map;
+
+public record KakaoOAuthResponse(Map<String, Object> attribute) {
+
+    public KakaoOAuthResponse(Map<String, Object> attribute) {
+        this.attribute = (Map<String, Object>) attribute.get("kakao_account");
+    }
+
+    public String getProvider() {
+        return "kakao";
+    }
+
+    public String getEmail() {
+        return attribute.get("email").toString();
+    }
+
+    public String getNickname() {
+        return ((Map<String, Object>) attribute.get("profile")).get("nickname").toString();
+    }
+}
```
이 클래스가 필요한 이유가 무엇인가요?

suhyeon7497 replied:
이 클래스 또한 향후에 다른 oauth로그인들을 위해서 만든 것이었습니다. 하하..
각 provider마다 response 형식이 다르기 때문에 provider 별로 따로 만들어줘야 했기 때문에 만들었습니다.

sanghee0820 commented on src/main/java/team7/inplace/security/domain/User.java:
```diff
@@ -0,0 +1,48 @@
+        this.userType = userType;
+    }
+
+    protected User() {
+    }
```
@NoArgesConstructor(access = PROTECTED) 사용은 어떠신가요?

suhyeon7497 replied:
넵!

sanghee0820 commented on src/main/java/team7/inplace/security/persistence/UserRepository.java:
```diff
@@ -0,0 +1,11 @@
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    User findByUsername(String username);
```
Optional<User>이 맞을 듯 싶습니다!

suhyeon7497 replied:
넵!

sanghee0820 reviewed (changes requested):
코멘트 몇개 남겼습니다!

수정 부탁드립니다.

그리고 추가로 JWT 발급 로직은 없을까요?

suhyeon7497 committed:
[chore] jwt 관련 dependency 추가

jwt 토큰 발급을 위해 dependency를 추가 하였습니다.

관련 이슈: #2

suhyeon7497 committed:
[feat] jwt를 발급, 검증하는 util 클래스

jwt토큰을 발급하고 검증하는 JWTUtil 클래스를 작성하였습니다.

관련 이슈: #2

BaeJunH0 reviewed (approved):
화이팅입니다 승인할게요~

suhyeon7497 committed:
[refactor] code review 피드백 반영

코드 리뷰 피드백을 반영하였습니다.

관련 이슈: #2

suhyeon7497 committed:
[feat] : cookie를 사용해 token 반환

cookie를 통해 accessToken과 refreshToken을 반환하였습니다.
현재는 localhost:8080/successCookie로 redirect 하도록 했습니다.

관련 이슈: #2

suhyeon7497 committed:
Merge branch 'weekly/2' of https://github.com/kakao-tech-campus-2nd-step3/Team7_BE into feat/#2-authentication

sanghee0820 reviewed (approved):
Util은 아직 Component 등록 안하신걸까요?

suhyeon7497 commented:
Bean으로 등록했습니다!
